src/Annotations/AppKernel.php

```diff
@@ -0,0 +1,281 @@
+                new CompositeLimit(
+                    $app['limit._memory'],
+                    $app['limit._signals']
+                ),
+                $app['logger']
+            );
+        };
+
+        $this->app['limit.interactive'] = function (Application $app) {
+            return new LoggingLimit(
+                $app['limit._signals'],
+                $app['logger']
+            );
+        };
+
+        $this->app['hypothesis.guzzle'] = function (Application $app) {
+            // Create default HandlerStack
+            $stack = HandlerStack::create();
+            $logger = $app['logger'];
+            if ($app['debug']) {
+                $stack->push(
+                    Middleware::mapRequest(function ($request) use ($logger) {
+                        $logger->debug("Request performed in Guzzle Middleware: {$request->getUri()}.", ['request' => str($request)]);
+
+                        return $request;
+                    })
+                );
+                $stack->push(
+                    Middleware::mapResponse(function ($response) use ($logger) {
+                        $logger->debug('Response received in Guzzle Middleware.', ['response' => str($response)]);
+                    })
+                );
+            }
+
+            return new Client([
+                'base_uri' => $app['hypothesis']['api_url'],
+                'handler' => $stack,
+            ]);
+        };
+
+        $this->app['hypothesis.sdk'] = function (Application $app) {
+            $notifyingHttpClient = new HypothesisNotifyingHttpClient(
+                new HypothesisBatchingHttpClient(
+                    new HypothesisGuzzle6HttpClient(
+                        $app['hypothesis.guzzle']
+                    ),
+                    $app['api.requests_batch']
+                )
+            );
+            if ($app['debug']) {
+                $logger = $app['logger'];
+                $notifyingHttpClient->addRequestListener(function ($request) use ($logger) {
+                    $logger->debug("Request performed in NotifyingHttpClient: {$request->getUri()}");
+                });
+            }
+
+            $credentials = new Credentials(
+                $app['hypothesis']['client_id'],
+                $app['hypothesis']['secret_key'],
+                $app['hypothesis']['authority']
+            );
+
+            return new HypothesisApiSdk($notifyingHttpClient, $credentials);
+        };
+
+        $this->app['guzzle'] = function (Application $app) {
+            // Create default HandlerStack
+            $stack = HandlerStack::create();
+            $logger = $app['logger'];
+            if ($app['debug']) {
+                $stack->push(
+                    Middleware::mapRequest(function ($request) use ($logger) {
+                        $logger->debug("Request performed in Guzzle Middleware: {$request->getUri()}");
+
+                        return $request;
+                    })
+                );
```

Comment:
@giorgiosironi do you also want me to expand what we are logging here?

Reply:
best-effort, we can easily tune this later and the current level (`debug`) poses no risk